### ISSUE TITLE
Deduplication and compression can't be disabled depending on the workload

### ISSUE THREAD

**Issue by patrickr92** [dm-vdo/kvdo#35]
Hello there,
I observed some unexpected behavior when I use the VDO flags `--deduplication=disabled` or `--compression=disabled` or both. It seems that deduplication and/or compression still remain enabled for some workloads. I ran multiple workloads where `vdo status` still reported space savings although both stages where disabled. I ensured that the workloads do not contain any zero-blocks. To me it seems that this behavior only occurs in workloads that contain duplicates and the temporal distance between them played a critical role.

### The following workloads(WL) were tested:

1. Compressible only, which **worked as expected**. Generated by
```
  mkdir compOnly; cd compOnly
  for i in $(seq 0 1023); do python -c "import os; X=''.join([(os.urandom(1536)+'\xde\xad\xbe\xef'*640) for i in range(256)])[0:-1]; print X" > randComp$i; done
  cat randComp* > compOnly
```
2. Dedupable only, _contiguous_ random data duplicated on 1Gb boundary (pattern like ABCD...ABCD...). This also **worked as expected**. Generated by 
```
  mkdir dedupCont; cd dedupCont
  for i in $(seq 0 1023); do dd if=/dev/urandom of=random_$i bs=1M count=1 ; done
  cat random_* > dedupOnly_contiguous; cat random_* >> dedupOnly_contiguous
```

3. Dedupable only, _interlaced_ random data duplicated per 1Mb boundary (pattern like AABBCCDD...).  Here, savings were reported even if deduplication and/or compression were disabled. Generated by
```
  mkdir dedupInterlaced; cd dedupInterlaced
  for i in $(seq 0 1023); do dd if=/dev/urandom of=random_$i bs=1M count=1 ; done
  for i in random*; do cat $i >> dedupOnly_interlaced; cat $i >> dedupOnly_interlaced; done
```

4. Compressible and dedupable, _contiguous_ compressible data duplicated on 1 Gb boundary (ABCD...ABCD...). Like the dedupable contiguous workload, it **worked as expected**. Generated by
```
  mkdir dedupCompCont; cd dedupCompCont
  for i in $(seq 0 1023); do python -c "import os; X=''.join([(os.urandom(1536)+'\xde\xad\xbe\xef'*640) for i in range(256)])[0:-1]; print X" > randComp$i; done
  cat randComp* > dedup_comp_contiguous; cat randComp* >> dedup_comp_contiguous
```

5. Compressible and dedupable, _interlaced_ compressible data duplicated per 1Mb boundary (AABBCCDD...). Like for the interlaced dedupable only workload, savings can be observed in all dedupe/comp=on/off combinations. Generated by
```
  mkdir dedupCompInterlaced; cd dedupCompInterlaced
  for i in $(seq 0 1023); do python -c "import os; X=''.join([(os.urandom(1536)+'\xde\xad\xbe\xef'*640) for i in range(256)])[0:-1]; print X" > randComp$i; done
  for i in randComp*; do cat $i >> dedup_comp_interlaced; cat $i >> dedup_comp_interlaced; done
```

### Test setup
For each test, a fresh VDO volume was created using vdo & kvdo version 6.1.3.23 on Centos 7.9.2009. However, short tests under Centos 8 revealed similar behavior. The workloads were tested using _dd_, but the issue was also observable using _fio_.
```
vdo create --name=<vol_name> --device=<dev-name> --compression=enabled --deduplication=enabled
dd if=<WL-file> of=/dev/mapper/<vdo_vol> oflag=direct status=progress
```

### The following savings were reported:
|                              | dedupe,comp=on | Only comp=on | Only dedupe=on | dedupe,comp=off |
|------------------------------|:--------------:|:------------:|:--------------:|:---------------:|
| 1. compression only WL          |       49%      |      49%     |       0%       |        0%       |
| 2. dedupe only WL, contiguously |       50%      |      0%      |       50%      |        0%       |
| **3. dedupe only WL, interlaced**   |       50%      |      **46%**     |       50%      |       **43%**       |
| 4. dedupe+comp WL, contiguously |       74%      |      49%     |       50%      |        0%       |
| **5. dedupe+comp WL, interlaced**   |       **66%**      |     **68%**     |       50%      |      **45%**       |

I assume the behavior is somehow triggered by the deduplication stage which seems to still deduplicate blocks that appear within a certain window of locality (here 1Mb). However, there is also some effect on the saving whether deduplication is turned on or off as the saving may decrease like in WL 3 or increase as in WL 5.

**Comment by bgurney-rh** [comment]
I'm assuming that you're creating a VDO volume, and then creating a filesystem on that VDO volume.

(In other words, `vdo create...` followed by `mkfs.ext4 ...` or `mkfs.xfs...`)

There is a very good chance that the `mkfs` command that writes the filesystem is writing zero blocks to the device.  (For filesystem metadata, this is particularly important, to avoid the chance of mixing old metadata with new metadata)

While the workloads do not contain zero blocks, the metadata areas of the filesystem being used may have had zero blocks that were written at mkfs time.

**Comment by patrickr92** [comment]
Thanks for the the quick response. 

> I'm assuming that you're creating a VDO volume, and then creating a filesystem on that VDO volume.

No, I was actually using the raw VDO volume which was directly written to by dd to confirm that issue.

**Comment by rhawalsh** [comment]
Hi @patrickr92 Thank you for the post!

From one of our developers:

The way we disable deduplication is to disable UDS, the persistent hash index
we use to find blocks on disk that might have the same contents as a write
request. We used to have poor performance when writing many copies of the same
data concurrently. Our solution for that deduplicates all active write
requests against each other without going through the UDS index, which is why
you're seeing dedupe even with the index disabled. When we added that
mechanism, we didn't add a way to disable it.

Writing blocks that are all-zero is a special case and they have always
deduplicated without going through either mechanism.
